### PR TITLE
Changing the `CanonicalName` of `DetachedHead` to match gitexe

### PR DESCRIPTION
**Pull request by whoisj**
Solves @Therzok issue in #1104.

Consider this a starting point for discussion. :smirk:

### CODE REVIEW

**Comment by Therzok** [comment]
LGTM given the tests get fixed. :D

**Comment by whoisj** [comment]
Sure. I assumed you'd like it :smirk: But is everyone else ok with this? 

/cc @nulltoken @jamill

File: LibGit2Sharp/DetachedHead.cs [hunk 1]
```diff
@@ -3,9 +3,12 @@
     internal class DetachedHead : Branch
     {
         internal DetachedHead(Repository repo, Reference reference)
-            : base(repo, reference, "(no branch)")
-        {
-        }
+            : base(repo,
+                   reference,
```
Comment (nulltoken) [inline LibGit2Sharp/DetachedHead.cs]: A detached `HEAD` should always be a `DirectReference`. Could we also change this ctor to only accept this type?

File: LibGit2Sharp/DetachedHead.cs [hunk 2]
```diff
@@ -3,9 +3,12 @@
...
+                   reference == null || reference.TargetIdentifier == null || reference.TargetIdentifier.Length < 7
```
Comment (nulltoken) [inline LibGit2Sharp/DetachedHead.cs]: Have we got any test that leverage this code path?
When would `reference == null || reference.TargetIdentifier == null` be leveraged?

**Comment by carlosmn** [comment]
In order to match git, we shouldn't be looking at HEAD, but at its reflog. It checks for the latest 'checkout' message and uses that to know what to show, which could be a tag name or an id prefix.

**Comment by whoisj** [comment]
> In order to match git, we shouldn't be looking at HEAD, but at its reflog. It checks for the latest 'checkout' message and uses that to know what to show, which could be a tag name or an id prefix.

Interesting. This would force us to refactor the current object hierarchy, but that's the kind of thing we to do now (in prep for v1.0).

**Pull request closed without merging**